File: csharp/Dson/src/DsonHeader.cs
#region LICENSE

//  Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System.Collections.Generic;

#pragma warning disable CS1591
namespace Wjybxx.Dson;

/// <summary>
/// Dson对象头
/// </summary>
/// <typeparam name="TK">String或<see cref="FieldNumber"/></typeparam>
public class DsonHeader<TK> : AbstractDsonObject<TK>
{
    public DsonHeader()
        : base(DsonInternals.NewLinkedDictionary<TK>(2)) {
    }

    public DsonHeader(IDictionary<TK, DsonValue> valueMap)
        : base(DsonInternals.NewLinkedDictionary(valueMap)) {
    }

    public override DsonType DsonType => DsonType.Header;

    public override DsonHeader<TK> Append(TK key, DsonValue value) {
        return (DsonHeader<TK>)base.Append(key, value);
    }
}

/// <summary>
/// 定义DsonHeader常量
/// </summary>
public static class DsonHeaders
{
    // header常见属性名
    public const string NamesClassName = "clsName";
    public const string NamesCompClassName = "compClsName";
    public const string NamesClassId = "clsId";
    public const string NamesCompClassId = "compClsId";
    public const string NamesLocalId = "localId";
    public const string NamesTags = "tags";
    public const string NamesNamespace = "ns";

    public static readonly FieldNumber NumbersClassName = FieldNumber.OfLnumber(0);
    public static readonly FieldNumber NumbersCompClassName = FieldNumber.OfLnumber(1);
    public static readonly FieldNumber NumbersClassId = FieldNumber.OfLnumber(2);
    public static readonly FieldNumber NumbersCompClassId = FieldNumber.OfLnumber(3);
    public static readonly FieldNumber NumbersLocalId = FieldNumber.OfLnumber(4);
    public static readonly FieldNumber NumbersTags = FieldNumber.OfLnumber(5);
    public static readonly FieldNumber NumbersNamespace = FieldNumber.OfLnumber(6);
}

File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonHeader.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.dson.internal.ValuesPolicy;

import javax.annotation.Nonnull;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * 1.Header不可以再持有header，否则陷入死循环
 * 2.Header的结构应该是简单清晰的，可简单编解码
 *
 * @author wjybxx
 * date - 2023/5/27
 */
@SuppressWarnings("unused")
public class DsonHeader<K> extends AbstractDsonObject<K> {

    public DsonHeader() {
        super(new LinkedHashMap<>(4), ValuesPolicy.SOURCE);
    }

    public DsonHeader(Map<K, DsonValue> valueMap) {
        super(valueMap, ValuesPolicy.COPY);
    }

    private DsonHeader(Map<K, DsonValue> valueMap, ValuesPolicy policy) {
        super(valueMap, policy);
    }

    public static <K> DsonHeader<K> toImmutable(DsonHeader<K> src) {
        return new DsonHeader<>(src.valueMap, ValuesPolicy.IMMUTABLE);
    }

    public static <K> DsonHeader<K> empty() {
        return new DsonHeader<>(Map.of(), ValuesPolicy.IMMUTABLE);
    }

    //

    @Nonnull
    @Override
    public final DsonType getDsonType() {
        return DsonType.HEADER;
    }

    /** @return this */
    @Override
    public DsonHeader<K> append(K key, DsonValue value) {
        put(key, value);
        return this;
    }

    @Override
    public String toString() {
        return "DsonHeader{" +
                "valueMap=" + valueMap +
                '}';
    }
//

    // header常见属性名
    public static final String NAMES_CLASS_NAME = "clsName";
    public static final String NAMES_COMP_CLASS_NAME = "compClsName";
    public static final String NAMES_CLASS_ID = "clsId";
    public static final String NAMES_COMP_CLASS_ID = "compClsId";
    public static final String NAMES_LOCAL_ID = "localId";
    public static final String NAMES_TAGS = "tags";
    public static final String NAMES_NAMESPACE = "ns";

    public static final int NUMBERS_CLASS_NAME = DsonLites.makeFullNumberZeroIdep(0);
    public static final int NUMBERS_COMP_CLASS_NAME = DsonLites.makeFullNumberZeroIdep(1);
    public static final int NUMBERS_CLASS_ID = DsonLites.makeFullNumberZeroIdep(2);
    public static final int NUMBERS_COMP_CLASS_ID = DsonLites.makeFullNumberZeroIdep(3);
    public static final int NUMBERS_LOCAL_ID = DsonLites.makeFullNumberZeroIdep(4);
    public static final int NUMBERS_TAGS = DsonLites.makeFullNumberZeroIdep(5);
    public static final int NUMBERS_NAMESPACE = DsonLites.makeFullNumberZeroIdep(6);
}